lib/Map/TableStructure.js

```diff
@@ -503,6 +514,123 @@ TableStructure.prototype.getIdMapping = function() {
+        updatedColumns.push(new TableColumn(column.name, updatedColumnValuesArrays[columnNumber], column.getFullOptions()));
+        return updatedColumns;
+    }, []);
+    // activeTimeColumn is not tracked, but columns are, so change activeTimeColumn first.
+    this.activeTimeColumn = getColumnWithSameId(this.activeTimeColumn, updatedColumns);
+    this.columns = updatedColumns;
+};
+
+/**
+ * Replace specific rows in this table with rows in table2.
+ * Changes all the columns in one go, to avoid partial updates from tracked values.
+ * @param {TableStructure} table2 The table whose rows should replace this table's rows.
+ * @param {Object} replacementMap An object whose properties are {table 1 row number: table 2 row number}.
+ */
+TableStructure.prototype.replaceRows = function(table2, replacementMap) {
+    var updatedColumnValuesArrays = [];
+    for (var columnNumber = 0; columnNumber < table2.columns.length; columnNumber++) {
+        updatedColumnValuesArrays.push(this.columns[columnNumber].values);
+        for (var table1RowNumber in replacementMap) {
+            if (replacementMap.hasOwnProperty(table1RowNumber)) {
+                var table2RowNumber = replacementMap[table1RowNumber];
+                updatedColumnValuesArrays[columnNumber][table1RowNumber] = table2.columns[columnNumber].values[table2RowNumber];
+            }
+        }
+    }
+    var updatedColumns = this.columns.map((column, columnNumber) =>
+        new TableColumn(column.name, updatedColumnValuesArrays[columnNumber], column.getFullOptions())
+    );
+    this.columns = updatedColumns;
+};
+
+function getColumnWithSameId(column1, columns) {
+    if (defined(column1)) {
+        var matchingColumns = columns.filter(column => column.id === column1.id);
+        if (matchingColumns.length !== 1) {
+            throw new DeveloperError('Ambiguous column: ' + column1.name);
+        }
+        return matchingColumns[0];
+    }
+}
+
+/**
+ * Merges table2 into this table.
+ * Uses this.idColumnNames (and this.activeTimeColumn, if present) to identify matching rows.
+ * Changes all the columns in one go, to avoid partial updates from tracked values.
+ * @param  {TableStructure} table2 The table to merge into this one.
+ */
+TableStructure.prototype.merge = function(table2) {
+    if (!defined(this.idColumnNames)) {
+        throw new DeveloperError('Cannot merge tables without id columns.');
+    }
+    if (this.columns.length !== table2.columns.length) {
+        throw new DeveloperError('Cannot merge tables with different numbers of columns.');
+    }
+    var table1RowNumbersMap = this.getIdMapping();
+    var table2RowNumbersMap = table2.getIdMapping(this.idColumnNames);
+    var rowsFromTable2ToAppend = [];  // An array of row numbers.
+    var rowsToReplace = {};  // Properties are {table 1 row number: table 2 row number}.
+    var table2ActiveTimeColumn = getColumnWithSameId(this.activeTimeColumn, table2.columns);
+    for (var key in table2RowNumbersMap) {
```

Comment:
I'm having trouble following this loop - could you maybe rename `key` to `table2Id`? That's what it is right?